### PR TITLE
implement __assume for non-windows builds

### PR DESCRIPTION
**Pull request by wecing**
issue #12808

### CODE REVIEW

**Comment by jkotas** [comment]
Build break on Ubuntu arm:

```
/opt/code/src/inc/stgpool.h:225:24: error: the argument to '__builtin_assume' has side effects that will be discarded [-Werror,-Wassume]
```

File: src/pal/inc/pal.h [hunk 1]
```diff
@@ -156,9 +156,14 @@ extern "C" {
...
 
 #if !defined(_MSC_VER) || defined(SOURCE_FORMATTING)
-#define __assume(x) (void)0
 #define __annotation(x)
-#endif //!MSC_VER
+#endif // !MSC_VER || SOURCE_FORMATTING
+
+#if !(defined(_MSC_VER) || defined(__llvm__)) || defined(SOURCE_FORMATTING)
+#define __assume(x) (void)0
+#elif defined(__llvm__) // __llvm__ && !SOURCE_FORMATTING
```
Comment (danmosemsft) [inline src/pal/inc/pal.h]: Wouldn't #if case have been true in this case

**Comment by AndyAyersMS** [comment]
@wecing can you also verify when this is building cleanly that the optimization we expect to see in `MethodTable::GetManagedClassObjectIfExists` happens?

Related issue has a test program you can try. A before/after disassembly of `GetManagedClassObjectIfExists` would also be nice to see.

File: src/pal/inc/pal.h (hunk 1 above)
Comment (wecing) [reply src/pal/inc/pal.h]: No, here defined(__llvm__) is true, so !(defined(_MSC_VER) || defined(__llvm__)) evaluates to false.

**Comment by wecing** [comment]
Looks like both versions produce the same MethodTable::GetManagedClassObjectIfExists assembly (only tested with clang 3.9). Also the __builtin_assume instruction was introduced in clang 3.6, so for now it would only be enabled for arm crossbuilds.

**Comment by AndyAyersMS** [comment]
I wonder if maybe LTO is required to see the improvement.

Does implementing this impact the generated code anywhere else?

**Comment by wecing** [comment]
For LTO, I tried
```
if (CLR_CMAKE_PLATFORM_UNIX)
  # set(INTERPROCEDURAL_OPTIMIZATION ON)
  set(CMAKE_SHARED_LINKER_FLAGS "${CMAKE_SHARED_LINKER_FLAGS} -flto")
  add_compile_options(-flto)
```

But got:

```
[  1%] Building CXX object src/nativeresources/CMakeFiles/nativeresourcestring.dir/resourcestring.cpp.o
CMakeFiles/System.Globalization.Native.dir/calendarData.cpp.o: file not recognized: File format not recognized
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```

> Does implementing this impact the generated code anywhere else?

Not any that I am aware of, but I only checked a few places.

**Comment by janvorli** [comment]
@wecing LTO requires using the gold linker, the standard ld doesn't understand it. Moreover, the LLVMgold.so plugin needs to be used with it. See details here: http://llvm.org/docs/GoldPlugin.html

**Comment by jkotas** [comment]
Also see https://github.com/dotnet/coreclr/pull/10533

**Comment by wecing** [comment]
It looks like even if I passed -clang3.9, the compiler used is still 3.5. gen-build-sys-clang.sh is passed "3 9" as well. Will dig deeper into this once I get a chance.

**Comment by janvorli** [comment]
@wecing you need to delete the bin folder to ensure that cmake picks the new clang option.

**Comment by wecing** [comment]
Not sure why the "Tizen armel Cross Debug Build" failed; it seems to happen on IA64 only.

@AndyAyersMS `MethodTable::GetManagedClassObjectIfExists` is `FORCEINLINE`. So I guess LTO doesn't really matter here?

**Comment by karelz** [comment]
What's the plan with this PR @RussKeldorph ? Maybe we should close it, until it is ready? There hasn't been any update for 1.5 years ...

**Pull request closed without merging**